### PR TITLE
Quantization folding pass

### PR DESCRIPTION
**Pull request by per**
### Summary
Adds a folding pass to fold in q and dq nodes.

### Test plan
Added test for the new pass

### CODE REVIEW

**Comment by per** [comment]
- linux-job failing due to 404 when downloading https://sdk.lunarg.com/sdk/download/1.2.198.1/linux/vulkansdk-linux-x86_64-1.2.198.1.tar.gz
- flaky test in Arm unittests not picking up `@pytest.mark.flaky` (not sure if flaky is installed) and not rerunning automatically.

File: backends/arm/_passes/fold_qdq_with_annotated_qparams_pass.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        output_dq: "f32[5]" = quantized_decomposed_dequantize_per_tensor_default(aten_add_tensor_q, 0.05487706884741783, -128, -128, 127, torch.int8)
+
+    The quantization parameters for x_dq and aten_add_tensor_q are store in meta for the aten_add_tensor node.
+
+    """
+
+    def __init__(self, targeted_ops: Iterable[Callable]):
+        super().__init__()
+        self.targeted_ops = targeted_ops
```
[the remark this answers is not in the document]
Comment (per) [reply backends/arm/_passes/fold_qdq_with_annotated_qparams_pass.py:51]: Yes, that's the plan. We would like to move all quantized operators to be handled this way.

File: backends/arm/test/passes/test_fold_qdq_pass.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from executorch.backends.arm.test.tester.arm_tester import ArmTester
+
+from executorch.backends.xnnpack.test.tester.tester import RunPasses
+
+from executorch.exir.dialects._ops import ops as exir_ops
+
+
+class SimpleQuantizeModel(torch.nn.Module):
+    def forward(self, x):
+        return x + x
```
[the remark this answers is not in the document]
Comment (per) [reply backends/arm/test/passes/test_fold_qdq_pass.py:24]: Ack.

File: backends/arm/operators/op_max.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            assert (
+                x_scale == y_scale
+            ), "Different scale for input, MAX should be quantized with shared quantization!"
+
+            operand_inputs, scale_back = tqutils.insert_rescale_ops_to_int32(
+                tosa_graph, inputs, node
+            )
+
+            output.shape = tosa_shape(output.shape, output.dim_order)
+            min_output = tosa_graph.addIntermediate(output.shape, ts.DType.INT32)
```
[the remark this answers is not in the document]
Comment (per) [reply backends/arm/operators/op_max.py:64]: Yepp!

File: backends/arm/operators/op_max.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+            y_scale = input_qparams[1].scale
+            y_zp = input_qparams[1].zp
+
+            assert (
+                x_zp == y_zp
+            ), "Different zp for inputs, MAX should be quantized with shared quantization!"
+            assert (
+                x_scale == y_scale
+            ), "Different scale for input, MAX should be quantized with shared quantization!"
```
[the remark this answers is not in the document]
Comment (per) [reply backends/arm/operators/op_max.py:57]: Yes will fix it up.

File: backends/arm/tosa_quant_utils.py [hunk 5]
```diff
@@ -42,6 +42,81 @@ def register_passable_op(op):
     passable_ops.append(op)
 
 
+def insert_rescale_ops_to_int32(
```
[the remark this answers is not in the document]
Comment (per) [reply backends/arm/tosa_quant_utils.py:45]: I'll keep the _ops_ naming, since the function inserts ops in the TOSA graph.

**Comment by per** [comment]
pull / unittest / macos / macos-job (pull_request)  failing seems to be unrelated (test_flamingo_vision_encoder)

**Review by digantdesai — APPROVED** [review]

**Pull request merged**